conbench/api/runs.py

```diff
@@ -145,16 +133,33 @@ def get(self):
         """
         if sha_arg := f.request.args.get("sha"):
             shas = sha_arg.split(",")
-            runs = Run.search(filters=[Commit.sha.in_(shas)], joins=[Commit])
+            filters = [Commit.sha.in_(shas)]
+            joins = [Commit]
         else:
-            runs = Run.all(order_by=Run.timestamp.desc(), limit=1000)
-        return self.serializer.many.dump(runs)
+            filters = []
+            joins = None
+
+        results = BenchmarkResult.search(
+            filters=filters,
+            joins=joins,
+            order_by=BenchmarkResult.timestamp.desc(),
+            limit=1000,
+        )
```

Comment:
A more generalized way to do this would be something like 

``` sql
SELECT *
FROM (
  SELECT 
    *, 
    ROW_NUMBER() OVER (PARTITION BY "run_id") AS "_row_index"
  FROM "result"
) "subquery1"
WHERE ("_row_index" = 1)
```

(Or selecting down and using `DISTINCT`, but that relies on results having identical run metadata.) It would surely be slower than the BMRT cache, but it would let us `LIMIT`/`OFFSET` paginate over everything reliably. 

I'm not sure it's the implementation we want now, but it seems worth running some timings on it to see what the cost of denormalization is without workarounds.

Reply:
@alistaire47 Yeah I considered something like that, but even on the Velox DB (which has at least an order of magnitude fewer benchmark results than Arrow) this query takes 45 seconds. Using the cache will be almost instantaneous. The only downside will be potential inaccuracies around the boundaries. But I think that's okay for our purposes. The cache refresh job runs every 2 minutes.

---

@jgehrcke in #1338 you decided to only populate the cache with results from the default branch. What, in your opinion, will be the fallout if I change this to allow results on all branches? Is it just performance considerations or will we have to adjust the c-bench UI code as well?

I'd like the `GET /api/runs/` endpoint and the Runs UI page to return runs that weren't on the default branch too.

Now I'm kind of thinking this is a blocker to using the BMRT cache for this purpose, so if you have some time in your morning and you want to make a quick patch outside of this PR, go for it. Else I will look into it later.

Reply:
> Yeah I considered something like that, but even on the Velox DB (which has at least an order of magnitude fewer benchmark results than Arrow) this query takes 45 seconds.

This has me concerned; I'm not sure we'll always be able to get around these types of queries as we take next steps like denormalizing hardware, querying for fingerprints, etc., especially if we want/need completeness for paginated APIs. I don't see easy solutions, but am worried about dragons in our unknowns. Am I repeating myself? I feel like I'm repeating myself, sorry.

Reply:
> What, in your opinion, will be the fallout if I change this to allow results on all branches?

I made this change (https://github.com/conbench/conbench/issues/1307, https://github.com/conbench/conbench/pull/1308) so that the plots (the 'tinyplots') contain data points only for the default branch. This can of course be achieved with other means, too. Let me think ...

Reply:
We talked in-person and decided to power these use cases using a variant of @alistaire47 's query that filters to all benchmark results from the last 30 days, no matter how many or few there are. This takes ~1 second on the Velox database. (I think this paradigm can set us up better for time-based pagination in the future anyway.)

Reply:
> time-based pagination

oh that's an interesting idea

Reply:
Getting back to this; https://github.com/conbench/conbench/issues/1466.